### DISCUSSION TITLE
Is it possible to adapt Connector to ADO.NET classes?

### DISCUSSION THREAD

**Discussion by AntiTenzor** [usinesoft/Cachalot#24] [Ideas]
Since there is a lot of ORM systems over plain old ADO.NET with its general
classes DBConnection and DBCommand,
"ADO.NET-compatible" connector will add many usecases around Cachalot.

How many time can it take to implement adapter class for this ecosystem?

P.S. Thank you very much for interesting project.

**Comment by usinesoft** [comment]
Hello, 
Thank you for your interest. It is an interesting idea.
I am right now preparing the release of a web monitoring tool. 

![image](https://user-images.githubusercontent.com/12799638/203989927-cab10eab-afa2-4659-b755-9f824082cbe1.png)

I will make a doodle in order to prioritize different user requests

**Reply by AntiTenzor** [reply]
> Hello, Thank you for your interest. It is an interesting idea. I am right now preparing the release of a web monitoring tool.
> 
> I will make a doodle in order to prioritize different user requests

Actually I'm looking at Cachalot for some time already and for all these years
I do not see a sweet point to put it in my system arch. ( I'm sorry, I'm a little bit silly sometimes... =(   )

Basically what I have is core **RDB** -- plain old relational database (a few of them actually) all of them
have ADO.NET fabric and all other convinient things.

And user-side **APP** which load some information from DB at start and make some additional queries
from time to time.

**I'd like to cache part of RDB to _Cachalot_ at the same machine where APP sits,
and to use it as super-fast in-memory read-only cache.**



**Idea 1)** To make this super easy and convienient, ideally there should be single Cachalot "master" instance.
I can populate it from RDB with some special utility.

But to make things as automatic as possible, it would be great, if other Cachalot instances can work in "slave" mode and to get all changes from its "master" autoMAGICally. ;-)


**Idea 2)** And to make communication between **APP** and **Cachalot** "slave" as common as possible,
it would be great to use connection objects, that mimics ADO.NET .


**Idea 3)** I can even imagine one more idea.
To make **Cachalot** a transparent SQL proxy.

**APP** sends SQL query to **Cachalot** instance.
Then it returns results either from memory cache,
or routs these queries to other backing RDB.

Such configuration makes repeating queries faster,
better utilizes DB connection pool and even hides connection strings
from end-user app, because actual connStrings now can be hidden inside a Cachalot box...


I know this is too much.
It looks like I already have a little bit of Christmas mood. =D

**Comment by usinesoft** [comment]
Hello, and I am sorry for the delay in answering you.
Here is my opinion on why mimicking a SQL server (or even making it a transparent SQL proxy) is not a good idea.
First, what we store are not flat objects like the lines in a classic table but full hierarchical documents that can contain collections and child objects. Of course, a line in a table is a  simpler case of a document so it can be stored. And in the last version, I added even a special layout to store “flat” objects very efficiently.
But an SQL database is hardly useful without JOIN operations. And JOIN operations are close to impossible to implement in a distributed system. The two instances that match a join should be collocated on the same node. With the current sharding algorithm (we chose the node as the hash of the primary key modulo the number of nodes) it is impossible.
I gave it a lot of thought and the only theoretical solution is to have a flexible partitioning key. It will be the primary key for one of the collections and the foreign key for the second one.
This implies designing the database from the beginning for a specific join which in my opinion is not a good idea.
And in the end, NO SQL databases mostly free us from the need to make JOIN operations and to map complex objects to a relational schema.
